crates/cubecl-opt/src/passes/integer_range_analysis.rs

```diff
@@ -0,0 +1,263 @@
+                                changes.inc();
+                            }
+                        }
+                    }
+                    Operator::Sub(binop) if binop.out.item().elem().is_int() => {
+                        if let Some(out_id) = var_id(&binop.out) {
+                            let lhs_range = range_of(opt, &binop.lhs);
+                            let rhs_range = range_of(opt, &binop.rhs);
+                            let out_range = lhs_range - rhs_range;
+                            if Some(&out_range) != opt.program.int_ranges.get(&out_id) {
+                                opt.program.int_ranges.insert(out_id, out_range);
+                                changes.inc();
+                            }
+                        }
+                    }
+                    Operator::Mul(binop) if binop.out.item().elem().is_int() => {
+                        if let Some(out_id) = var_id(&binop.out) {
+                            let lhs_range = range_of(opt, &binop.lhs);
+                            let rhs_range = range_of(opt, &binop.rhs);
+                            let out_range = lhs_range * rhs_range;
+                            if Some(&out_range) != opt.program.int_ranges.get(&out_id) {
+                                opt.program.int_ranges.insert(out_id, out_range);
+                                changes.inc();
+                            }
+                        }
+                    }
+                    Operator::Div(binop) if binop.out.item().elem().is_int() => {
+                        if let Some(out_id) = var_id(&binop.out) {
+                            let lhs_range = range_of(opt, &binop.lhs);
+                            let rhs_range: Range = range_of(opt, &binop.rhs);
+                            let out_range = lhs_range / rhs_range;
+                            if Some(&out_range) != opt.program.int_ranges.get(&out_id) {
+                                opt.program.int_ranges.insert(out_id, out_range);
+                                changes.inc();
+                            }
+                        }
+                    }
+                    Operator::Modulo(binop) if binop.out.item().elem().is_int() => {
+                        if let Some(out_id) = var_id(&binop.out) {
+                            let lhs_range = range_of(opt, &binop.lhs);
+                            let rhs_range = range_of(opt, &binop.rhs);
+                            let out_range = lhs_range % rhs_range;
+                            if Some(&out_range) != opt.program.int_ranges.get(&out_id) {
+                                opt.program.int_ranges.insert(out_id, out_range);
+                                changes.inc();
+                            }
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+}
+
+pub(crate) fn range_of(opt: &Optimizer, var: &Variable) -> Range {
+    match var {
+        Variable::Versioned {
```

Comment:
What is a versioned variable? Is it only for the optimization?

Reply:
Yes, it's a local variable that has been transformed into an immutable by versioning it with each assignment. It should be treated the same as a `LocalBinding`, but needs the extra version number to make SSA transformation possible (since we don't have the `CubeContext` to allocate new binding IDs when we get to the optimizer, so we have to instead version).